python/cudf/cudf/core/udf/lambda_function.py

```diff
@@ -0,0 +1,91 @@
+from numba import cuda
+from numba.np import numpy_support
+
+from cudf.core.udf.api import Masked, pack_return
+from cudf.core.udf.templates import (
+    lambda_kernel_template,
+    masked_input_initializer_template,
+    unmasked_input_initializer_template,
+)
+from cudf.core.udf.typing import MaskedType
+from cudf.core.udf.utils import (
+    construct_signature,
+    get_udf_return_type,
+    mask_get,
+)
+
+
+def lambda_kernel_from_template(sr, args):
+    """
+    Function to write numba kernels for `Series.apply` as a string.
+    Workaround until numba supports functions that use `*args`
+
+    `Series.apply` expects functions of a single variable and possibly
+    one or more constants, such as:
+
+    def f(x, c, k):
+        return (x + c) / k
+
+    where the `x` are meant to be the values of the series. Since there
+    can be only one column, the only thing that varies in the kinds of
+    kernels that we want is the number of extra_args
+
+    def _kernel(retval,
+                size,
+                input_col_0,
+                offset_0,
+                extra_arg_0, # the extra arg `c`
+                extra_arg_1, # the extra arg `k`
+    ):
+        i = cuda.grid(1)
+        ret_data_arr, ret_mask_arr = retval
+
+        if i < size:
+            d_0, m_0 = input_col_0
+            masked_0 = Masked(d_0[i], mask_get(m_0, i + offset_0)
+
+            ret = f_(masked_0, extra_arg_0, extra_arg_1)
+
+            ret_masked = pack_return(ret)
+            ret_data_arr[i] = ret_masked.value
+            ret_mask_arr[i] = ret_masked.valid
+
+    """
+    extra_args = ", ".join([f"extra_arg_{i}" for i in range(len(args))])
+
+    masked_initializer = (
+        masked_input_initializer_template
+        if sr._column.mask
+        else unmasked_input_initializer_template
+    )
+    masked_initializer = masked_initializer.format(idx=0)
+
+    return lambda_kernel_template.format(
+        extra_args=extra_args, masked_initializer=masked_initializer
+    )
+
+
+def get_lambda_kernel(sr, func, args):
```

Comment:
It seems like this function is very similar to `get_row_kernel` and rather than two separate functions we could just write a common function that accepts a couple of extra arguments such as the data types for the rows. Would that be feasible?

Reply:
If we can be somewhat certain that the udfs supported by pandas series and dataframe will be of similar signature in the future then this would be a good idea. Otherwise I would suggest keeping them separate to decouple them.

Reply:
The reason I hesitate to have one top level function here is that at some point series and data frame functions need to be handled differently and if we only had one top level function, we’d need to detect which situation we’re in somewhere else, probably inside `indexed_frame.py` which seems worse to me. That said there’s definitely at least some common stuff that can be factored out here, which I will do.